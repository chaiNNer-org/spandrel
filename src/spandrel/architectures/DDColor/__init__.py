from typing import Literal

import torch
import torch.nn.functional as F  # noqa: N812
from torch import Tensor

from ...__helpers.model_descriptor import (
    ImageModelDescriptor,
    ModelTiling,
    StateDict,
)
from ..__arch_helpers.color import lab_to_rgb, linear_rgb_to_lab, rgb_to_linear_rgb
from ..__arch_helpers.state import get_seq_len
from .arch.ddcolor import DDColor


def _call(model: DDColor, input: Tensor) -> Tensor:
    if model.num_output_channels != 2:
        raise ValueError("The call API for DDColor only supports 2 output channels")

    # input tensor to have the shape (1, 1, H, W)
    if len(input.shape) != 4 or input.shape[0] != 1 or input.shape[1] != 1:
        raise ValueError(
            f"The call API for DDColor expected the input tensor to be of the shape (1, 1, H, W), but found {input.shape}"
        )

    h = input.shape[2]
    w = input.shape[3]

    # This implements the pipeline from the original DDColor code adopted to tensor format:
    # https://github.com/piddnad/DDColor/blob/e13093cd8a5c9667b32d358fb0c3473661e6031c/inference/colorization_pipline.py#L58-L82

    input_gray_linear = rgb_to_linear_rgb(input)
    # (1, 3, H, W)
    input_rgb_linear = torch.hstack([input_gray_linear] * 3)

    # (1, 1, H, W)
    orig_l = linear_rgb_to_lab(input_rgb_linear)[:, :1, :, :]

    # (1, 3, M, M) (M = model.input_size)
    img = F.interpolate(input_rgb_linear, size=model.input_size, mode="bilinear")
    # (1, 1, M, M)
    img_l = linear_rgb_to_lab(img)[:, :1, :, :]
    img_zero = torch.zeros_like(img_l)
    # (1, 3, M, M)
    img_gray_lab = torch.hstack([img_l, img_zero, img_zero])
    img_gray_rgb = lab_to_rgb(img_gray_lab)

    # (1, 2, M, M)
    output_ab = model(img_gray_rgb)

    # resize ab -> concat original l -> rgb
    # (1, 2, H, W)
    output_ab_resize = F.interpolate(output_ab, size=(h, w), mode="bilinear")
    output_lab = torch.hstack([orig_l, output_ab_resize])
    output_rgb = lab_to_rgb(output_lab)

    # (1, 3, H, W)
    return output_rgb


def _get_encoder_convnext_depths_and_dims(state_dict: StateDict):
    depths = [
        get_seq_len(state_dict, "encoder.arch.stages.0"),
        get_seq_len(state_dict, "encoder.arch.stages.1"),
        get_seq_len(state_dict, "encoder.arch.stages.2"),
        get_seq_len(state_dict, "encoder.arch.stages.3"),
    ]
    dims = [
        state_dict["encoder.arch.stages.0.0.norm.weight"].shape[0],
        state_dict["encoder.arch.stages.1.0.norm.weight"].shape[0],
        state_dict["encoder.arch.stages.2.0.norm.weight"].shape[0],
        state_dict["encoder.arch.stages.3.0.norm.weight"].shape[0],
    ]
    return depths, dims


def load(state_dict: StateDict) -> ImageModelDescriptor[DDColor]:
    # they commented out head_cls, so we have to clean it up
    state_dict.pop("encoder.arch.head_cls.weight", None)
    state_dict.pop("encoder.arch.head_cls.bias", None)

    # default values
    encoder_name = "convnext-l"
    decoder_name = "MultiScaleColorDecoder"
    num_input_channels = 3  # has to be 3 or else DDColor will error
    input_size = (256, 256)
    nf = 512
    num_output_channels = 3
    last_norm: Literal["Batch", "BatchZero", "Weight", "Spectral"] = "Weight"
    do_normalize = False  # cannot be deduced from state_dict
    num_queries = 256
    num_scales = 3
    dec_layers = 9

    num_output_channels = state_dict["refine_net.0.0.weight_orig"].shape[0]
    num_queries = state_dict["refine_net.0.0.weight_orig"].shape[1] - 3

    if "decoder.last_shuf.conv.1.weight" in state_dict:
        last_norm = "Batch"  # or "BatchZero", but we can't tell which
        nf = state_dict["decoder.last_shuf.conv.1.weight"].shape[0] // 8
    elif "decoder.last_shuf.conv.0.weight_orig" in state_dict:
        last_norm = "Spectral"
        nf = state_dict["decoder.last_shuf.conv.0.weight_orig"].shape[0] // 8
    elif "decoder.last_shuf.conv.0.weight_g" in state_dict:
        last_norm = "Weight"
        nf = state_dict["decoder.last_shuf.conv.0.weight_g"].shape[0] // 8
    else:
        raise ValueError("Unknown last norm")

    if "decoder.color_decoder.query_feat.weight" in state_dict:
        decoder_name = "MultiScaleColorDecoder"
        num_scales = state_dict["decoder.color_decoder.level_embed.weight"].shape[0]
        dec_layers = get_seq_len(
            state_dict, "decoder.color_decoder.transformer_self_attention_layers"
        )
    else:
        decoder_name = "SingleColorDecoder"
        # num_scales and dec_layers aren't defined for SingleColorDecoder,
        # so we just assume default values
        num_scales = 3
        dec_layers = 9

    depths, dims = _get_encoder_convnext_depths_and_dims(state_dict)
    if depths == [3, 3, 9, 3] and dims == [96, 192, 384, 768]:
        encoder_name = "convnext-t"
    elif depths == [3, 3, 27, 3] and dims == [96, 192, 384, 768]:
        encoder_name = "convnext-s"
    elif depths == [3, 3, 27, 3] and dims == [128, 256, 512, 1024]:
        encoder_name = "convnext-b"
    elif depths == [3, 3, 27, 3] and dims == [192, 384, 768, 1536]:
        encoder_name = "convnext-l"
    else:
        raise ValueError("Unknown encoder architecture")

    model = DDColor(
        encoder_name=encoder_name,
        decoder_name=decoder_name,
        num_input_channels=num_input_channels,
        input_size=input_size,
        nf=nf,
        num_output_channels=num_output_channels,
        last_norm=last_norm,
        do_normalize=do_normalize,
        num_queries=num_queries,
        num_scales=num_scales,
        dec_layers=dec_layers,
    )

    return ImageModelDescriptor(
        model,
        state_dict,
        architecture="DDColor",
        purpose="Restoration",
        tags=[
            decoder_name,
            last_norm,
            encoder_name,
            f"{nf}nf",
            f"{num_queries}nq {num_scales}ns {dec_layers}dl",
        ],
        supports_half=False,
        supports_bfloat16=True,
        scale=1,
        input_channels=1,
        output_channels=3,
        tiling=ModelTiling.INTERNAL,
        call_fn=_call,
    )
