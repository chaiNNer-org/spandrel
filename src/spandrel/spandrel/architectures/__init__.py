"""
The package containing the implementations of all supported architectures. Not necessary for most user code.
"""
