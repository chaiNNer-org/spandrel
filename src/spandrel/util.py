"""
A module containing commonly-used functionality to implement architectures.
"""

from __future__ import annotations

import functools
import inspect
import math
from typing import Any, Literal, Mapping, Protocol, TypeVar


class KeyCondition:
    """
    A condition that checks if a state dict has the given keys.
    """

    def __init__(
        self, kind: Literal["all", "any"], keys: tuple[str | KeyCondition, ...]
    ):
        self._keys = keys
        self._kind = kind

    @staticmethod
    def has_all(*keys: str | KeyCondition) -> KeyCondition:
        return KeyCondition("all", keys)

    @staticmethod
    def has_any(*keys: str | KeyCondition) -> KeyCondition:
        return KeyCondition("any", keys)

    def __call__(self, state_dict: Mapping[str, object]) -> bool:
        def _detect(key: str | KeyCondition) -> bool:
            if isinstance(key, KeyCondition):
                return key(state_dict)
            return key in state_dict

        if self._kind == "all":
            return all(_detect(key) for key in self._keys)
        else:
            return any(_detect(key) for key in self._keys)


def get_first_seq_index(state_dict: Mapping[str, object], key_pattern: str) -> int:
    """
    Returns the maximum index `i` such that `key_pattern.format(str(i))` is in `state`.

    If no such key is in state, then `-1` is returned.

    Example:
        get_first_seq_index(state, "body.{}.weight") -> -1
        get_first_seq_index(state, "body.{}.weight") -> 3
    """
    for i in range(100):
        if key_pattern.format(str(i)) in state_dict:
            return i
    return -1


def get_seq_len(state_dict: Mapping[str, object], seq_key: str) -> int:
    """
    Returns the length of a sequence in the state dict.

    The length is detected by finding the maximum index `i` such that
    `{seq_key}.{i}.{suffix}` is in `state` for some suffix.

    Example:
        get_seq_len(state, "body") -> 5
    """
    prefix = seq_key + "."

    keys: set[int] = set()
    for k in state_dict.keys():
        if k.startswith(prefix):
            index = k[len(prefix) :].split(".", maxsplit=1)[0]
            keys.add(int(index))

    if len(keys) == 0:
        return 0
    return max(keys) + 1


def get_scale_and_output_channels(x: int, input_channels: int) -> tuple[int, int]:
    """
    Returns a scale and number of output channels such that `scale**2 * out_nc = x`.

    This is commonly used for pixelshuffel layers.
    """
    # Unfortunately, we do not have enough information to determine both the scale and
    # number output channels correctly *in general*. However, we can make some
    # assumptions to make it good enough.
    #
    # What we know:
    # - x = scale * scale * output_channels
    # - output_channels is likely equal to input_channels
    # - output_channels and input_channels is likely 1, 3, or 4
    # - scale is likely 1, 2, 4, or 8

    def is_square(n: int) -> bool:
        return math.sqrt(n) == int(math.sqrt(n))

    # just try out a few candidates and see which ones fulfill the requirements
    candidates = [input_channels, 3, 4, 1]
    for c in candidates:
        if x % c == 0 and is_square(x // c):
            return int(math.sqrt(x // c)), c

    raise AssertionError(
        f"Expected output channels to be either 1, 3, or 4."
        f" Could not find a pair (scale, out_nc) such that `scale**2 * out_nc = {x}`"
    )


def store_hyperparameters(*, extra_parameters: Mapping[str, object] = {}):
    """
    Stores the hyperparameters of a class in a `hyperparameters` attribute.
    """

    def get_arg_defaults(spec: inspect.FullArgSpec) -> dict[str, Any]:
        defaults = {}
        if spec.kwonlydefaults is not None:
            defaults = spec.kwonlydefaults

        if spec.defaults is not None:
            defaults = {
                **defaults,
                **dict(zip(spec.args[-len(spec.defaults) :], spec.defaults)),
            }

        return defaults

    class WithHyperparameters(Protocol):
        hyperparameters: dict[str, Any]

    C = TypeVar("C", bound=WithHyperparameters)

    def inner(cls: type[C]) -> type[C]:
        old_init = cls.__init__

        spec = inspect.getfullargspec(old_init)
        defaults = get_arg_defaults(spec)

        if spec.varargs is not None:
            raise UserWarning(
                "Class has *args, which is not allowed in combination with @store_hyperparameters"
            )
        if spec.varkw is not None:
            raise UserWarning(
                "Class has **kwargs, which is not allowed in combination with @store_hyperparameters"
            )

        @functools.wraps(old_init)
        def new_init(self: C, **kwargs):
            # remove extra parameters from kwargs
            for k, v in extra_parameters.items():
                if k in kwargs:
                    if kwargs[k] != v:
                        raise ValueError(
                            f"Expected hyperparameter {k} to be {v}, but got {kwargs[k]}"
                        )
                    del kwargs[k]

            self.hyperparameters = {**extra_parameters, **defaults, **kwargs}
            old_init(self, **kwargs)

        cls.__init__ = new_init
        return cls

    return inner


__all__ = [
    "KeyCondition",
    "get_first_seq_index",
    "get_seq_len",
    "get_scale_and_output_channels",
    "store_hyperparameters",
]
