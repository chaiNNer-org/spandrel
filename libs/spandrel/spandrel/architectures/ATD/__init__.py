import math

from typing_extensions import override

from spandrel.util import KeyCondition, get_pixelshuffle_params, get_seq_len

from ...__helpers.model_descriptor import (
    Architecture,
    ImageModelDescriptor,
    SizeRequirements,
    StateDict,
)
from .arch.atd_arch import ATD


class ATDArch(Architecture[ATD]):
    def __init__(
        self,
    ) -> None:
        super().__init__(
            id="ATD",
            name="ATD",
            detect=KeyCondition.has_all(
                "relative_position_index_SA",
                "conv_first.weight",
                "conv_first.bias",
                "layers.0.residual_group.td",
                "layers.0.residual_group.layers.0.sigma",
                "layers.0.residual_group.layers.0.norm1.weight",
                "layers.0.residual_group.layers.0.norm1.bias",
                "layers.0.residual_group.layers.0.norm2.weight",
                "layers.0.residual_group.layers.0.norm2.bias",
                "layers.0.residual_group.layers.0.norm3.weight",
                "layers.0.residual_group.layers.0.norm3.bias",
                "layers.0.residual_group.layers.0.wqkv.weight",
                "layers.0.residual_group.layers.0.attn_win.relative_position_bias_table",
                "layers.0.residual_group.layers.0.attn_win.proj.weight",
                "layers.0.residual_group.layers.0.attn_win.proj.bias",
                "layers.0.residual_group.layers.0.attn_atd.scale",
                "layers.0.residual_group.layers.0.attn_atd.wq.weight",
                "layers.0.residual_group.layers.0.attn_atd.wk.weight",
                "layers.0.residual_group.layers.0.attn_atd.wv.weight",
                "layers.0.residual_group.layers.0.attn_aca.logit_scale",
                "layers.0.residual_group.layers.0.attn_aca.proj.weight",
                "layers.0.residual_group.layers.0.convffn.fc1.weight",
                "layers.0.residual_group.layers.0.convffn.fc1.bias",
                "layers.0.residual_group.layers.0.convffn.dwconv.depthwise_conv.0.weight",
                "layers.0.residual_group.layers.0.convffn.dwconv.depthwise_conv.0.bias",
                "layers.0.residual_group.layers.0.convffn.fc2.weight",
                "layers.0.residual_group.layers.0.convffn.fc2.bias",
                "norm.weight",
                "norm.bias",
            ),
        )

    @override
    def load(self, state_dict: StateDict) -> ImageModelDescriptor[ATD]:
        img_size = 64  # cannot be deduced from state dict
        patch_size = 1  # cannot be deduced from state dict
        in_chans = 3
        embed_dim = 90
        depths = (6, 6, 6, 6)
        num_heads = (6, 6, 6, 6)
        window_size = 8
        category_size = 256  # cannot be deduced from state dict
        num_tokens = 64
        reducted_dim = 4
        convffn_kernel_size = 5
        mlp_ratio = 2.0
        qkv_bias = True
        ape = False
        patch_norm = True
        upscale = 1
        img_range = 1.0  # cannot be deduced from state dict
        upsampler = ""
        resi_connection = "1conv"
        norm = True

        in_chans = state_dict["conv_first.weight"].shape[1]
        embed_dim = state_dict["conv_first.weight"].shape[0]

        window_size = math.isqrt(state_dict["relative_position_index_SA"].shape[0])

        num_layers = get_seq_len(state_dict, "layers")
        depths = [6] * num_layers
        num_heads = [6] * num_layers
        for i in range(num_layers):
            depths[i] = get_seq_len(state_dict, f"layers.{i}.residual_group.layers")
            num_heads[i] = state_dict[
                f"layers.{i}.residual_group.layers.0.attn_win.relative_position_bias_table"
            ].shape[1]

        num_tokens = state_dict[
            "layers.0.residual_group.layers.0.attn_atd.scale"
        ].shape[0]
        reducted_dim = state_dict[
            "layers.0.residual_group.layers.0.attn_atd.wq.weight"
        ].shape[0]
        convffn_kernel_size = state_dict[
            "layers.0.residual_group.layers.0.convffn.dwconv.depthwise_conv.0.weight"
        ].shape[2]
        mlp_ratio = (
            state_dict["layers.0.residual_group.layers.0.convffn.fc1.weight"].shape[0]
            / embed_dim
        )
        qkv_bias = "layers.0.residual_group.layers.0.wqkv.bias" in state_dict
        ape = "absolute_pos_embed" in state_dict
        patch_norm = "patch_embed.norm.weight" in state_dict

        resi_connection = "1conv" if "layers.0.conv.weight" in state_dict else "3conv"

        if "conv_up1.weight" in state_dict:
            upsampler = "nearest+conv"
            upscale = 4
        elif "conv_before_upsample.0.weight" in state_dict:
            upsampler = "pixelshuffle"
            upscale, _ = get_pixelshuffle_params(state_dict, "upsample")
        elif "conv_last.weight" in state_dict:
            upsampler = ""
            upscale = 1
        else:
            upsampler = "pixelshuffledirect"
            upscale = math.isqrt(state_dict["upsample.0.weight"].shape[0] // in_chans)

        norm = "no_norm" not in state_dict

        is_light = upsampler == "pixelshuffledirect" and embed_dim == 48
        # use a heuristic for category_size
        category_size = 128 if is_light else 256

        tags = [f"{embed_dim}dim", f"{window_size}w", f"{category_size}cat"]
        if is_light:
            tags.insert(0, "light")

        model = ATD(
            img_size=img_size,
            patch_size=patch_size,
            in_chans=in_chans,
            embed_dim=embed_dim,
            depths=depths,
            num_heads=num_heads,
            window_size=window_size,
            category_size=category_size,
            num_tokens=num_tokens,
            reducted_dim=reducted_dim,
            convffn_kernel_size=convffn_kernel_size,
            mlp_ratio=mlp_ratio,
            qkv_bias=qkv_bias,
            ape=ape,
            patch_norm=patch_norm,
            upscale=upscale,
            img_range=img_range,
            upsampler=upsampler,
            resi_connection=resi_connection,
            norm=norm,
        )

        return ImageModelDescriptor(
            model,
            state_dict,
            architecture=self,
            purpose="Restoration" if upscale == 1 else "SR",
            tags=tags,
            supports_half=False,  # TODO: verify
            supports_bfloat16=True,
            scale=upscale,
            input_channels=in_chans,
            output_channels=in_chans,
            size_requirements=SizeRequirements(minimum=8),
        )
