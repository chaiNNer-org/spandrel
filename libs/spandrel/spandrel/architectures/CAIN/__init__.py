from __future__ import annotations

import math
from typing import Union

from typing_extensions import override

from spandrel.util import KeyCondition

from ...__helpers.model_descriptor import Architecture, ImageModelDescriptor, StateDict
from .arch.cain import CAIN
from .arch.cain_encdec import CAIN_EncDec
from .arch.cain_noca import CAIN_NoCA

_CainModel = Union[CAIN, CAIN_NoCA, CAIN_EncDec]


class CAINArch(Architecture[_CainModel]):
    def __init__(self) -> None:
        super().__init__(
            id="CAIN",
            detect=KeyCondition.has_any(
                # CAIN
                KeyCondition.has_all(
                    "encoder.interpolate.headConv.weight",
                    "encoder.interpolate.headConv.bias",
                    "encoder.interpolate.body.0.body.0.body.0.conv.weight",
                    "encoder.interpolate.body.0.body.0.body.2.conv.weight",
                    "encoder.interpolate.body.0.body.0.body.3.conv_du.0.weight",
                    "encoder.interpolate.body.0.body.1.body.0.conv.weight",
                    "encoder.interpolate.body.0.body.1.body.3.conv_du.2.weight",
                    "encoder.interpolate.body.0.body.3.body.2.conv.weight",
                    "encoder.interpolate.body.3.body.8.body.0.conv.weight",
                    "encoder.interpolate.body.4.body.5.body.3.conv_du.2.weight",
                    "encoder.interpolate.tailConv.weight",
                    "encoder.interpolate.tailConv.bias",
                ),
                # CAIN_NoCA
                KeyCondition.has_all(
                    "encoder.interpolate.headConv.weight",
                    "encoder.interpolate.headConv.bias",
                    "encoder.interpolate.body.0.body.0.body.0.conv.weight",
                    "encoder.interpolate.body.0.body.0.body.2.conv.weight",
                    "encoder.interpolate.body.0.body.1.body.0.conv.weight",
                    "encoder.interpolate.body.0.body.3.body.2.conv.weight",
                    "encoder.interpolate.body.3.body.8.body.0.conv.weight",
                    "encoder.interpolate.tailConv.weight",
                    "encoder.interpolate.tailConv.bias",
                ),
                # CAIN_EncDec
                KeyCondition.has_all(
                    "encoder.body.0.conv.weight",
                    "encoder.body.6.conv.weight",
                    "encoder.interpolate.headConv.weight",
                    "encoder.interpolate.headConv.bias",
                    "encoder.interpolate.body.0.body.0.body.0.conv.weight",
                    "encoder.interpolate.body.0.body.0.body.2.conv.weight",
                    "encoder.interpolate.body.0.body.0.body.3.conv_du.0.weight",
                    "encoder.interpolate.body.0.body.1.body.0.conv.weight",
                    "encoder.interpolate.body.0.body.1.body.3.conv_du.2.weight",
                    "encoder.interpolate.body.0.body.3.body.2.conv.weight",
                    "encoder.interpolate.body.3.body.8.body.0.conv.weight",
                    "encoder.interpolate.body.4.body.5.body.3.conv_du.2.weight",
                    "encoder.interpolate.tailConv.weight",
                    "encoder.interpolate.tailConv.bias",
                    "decoder.body.1.body.0.conv.weight",
                    "decoder.body.1.body.2.conv.bias",
                    "decoder.body.3.body.0.conv.weight",
                    "decoder.body.3.body.2.conv.bias",
                    "decoder.body.5.body.0.conv.weight",
                    "decoder.body.5.body.2.conv.bias",
                ),
            ),
        )

    @override
    def load(self, state_dict: StateDict) -> ImageModelDescriptor[_CainModel]:
        tags: list[str] = []

        if "encoder.body.0.conv.weight" in state_dict:
            start_filters = 32
            up_mode = "shuffle"

            start_filters = state_dict["decoder.body.5.body.0.conv.weight"].shape[0]

            if "decoder.body.0.upconv.0.conv.weight" in state_dict:
                up_mode = "shuffle"
            elif "decoder.body.0.upconv.weight" in state_dict:
                up_mode = "transpose"
            else:
                up_mode = "direct"

            model = CAIN_EncDec(
                start_filters=start_filters,
                up_mode=up_mode,
            )
            tags.extend(("EncDec", f"{start_filters}sf"))
        else:
            depth = 3

            # detect
            # n_feats = 3 * (4**depth)
            n_feats = state_dict["encoder.interpolate.headConv.weight"].shape[1]
            depth = math.isqrt(math.isqrt(n_feats // 3))

            if (
                "encoder.interpolate.body.0.body.0.body.3.conv_du.0.weight"
                in state_dict
            ):
                model = CAIN(
                    depth=depth,
                )
            else:
                model = CAIN_NoCA(
                    depth=depth,
                )
                tags.append("NoCA")

            tags.append(f"{depth}depth")

        return ImageModelDescriptor(
            model,
            state_dict,
            architecture=self,
            purpose="Restoration",
            tags=tags,
            supports_half=True,
            supports_bfloat16=True,
            scale=1,
            input_channels=3,
            output_channels=3,
        )
