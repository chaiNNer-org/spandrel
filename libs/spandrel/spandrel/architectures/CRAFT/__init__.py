import math

from typing_extensions import override

from spandrel.util import KeyCondition, get_seq_len

from ...__helpers.model_descriptor import (
    Architecture,
    ImageModelDescriptor,
    SizeRequirements,
    StateDict,
)
from .arch.CRAFT import CRAFT


class CRAFTArch(Architecture[CRAFT]):
    def __init__(
        self,
    ) -> None:
        super().__init__(
            id="CRAFT",
            detect=KeyCondition.has_all(
                "conv_first.weight",
                "layers.0.residual_group.hf_blocks.0.attn.temperature",
            ),
        )

    @override
    def load(self, state_dict: StateDict) -> ImageModelDescriptor[CRAFT]:
        # default values
        in_chans = 3
        window_size = 16  # cannot be deduced from state dict
        embed_dim = 48
        depths = [2, 2, 2, 2]
        num_heads = [6, 6, 6, 6]
        split_size_0 = 4
        split_size_1 = 16
        mlp_ratio = 2.0
        qkv_bias = True
        qk_scale = None  # cannot be deduced from state dict
        upscale = 4
        img_range = 1.0
        resi_connection = "1conv"

        # detect parameters
        in_chans = state_dict["conv_first.weight"].shape[1]
        embed_dim = state_dict["conv_first.weight"].shape[0]

        num_layers = get_seq_len(state_dict, "layers")
        depths = [2] * num_layers
        num_heads = [6] * num_layers
        for i in range(num_layers):
            depths[i] = get_seq_len(state_dict, f"layers.{i}.residual_group.hf_blocks")
            num_heads[i] = state_dict[
                f"layers.{i}.residual_group.hf_blocks.0.attn.temperature"
            ].shape[0]

        upscale = int(math.sqrt(state_dict["upsample.0.bias"].shape[0] / in_chans))

        if "conv_after_body.weight" in state_dict:
            resi_connection = "1conv"
        else:
            resi_connection = "identity"

        qkv_bias = "layers.0.residual_group.srwa_blocks.0.qkv.bias" in state_dict

        mlp_hidden_dim = state_dict[
            "layers.0.residual_group.srwa_blocks.0.mlp.fc1.bias"
        ].shape[0]
        mlp_ratio = float(mlp_hidden_dim / embed_dim)

        # Now split_size_0 and split_size_1:
        # What we know:
        #   a = s0 * s1
        #   b = (2*s0-1) * (2*s1-1)
        a = state_dict["relative_position_index_h"].shape[0]
        b = state_dict["biases_v"].shape[0]
        # Let's rearrange:
        #   s0 = a / s1
        #   b = (2*a/s1-1) * (2*s1-1)
        # Solve for s1 (wolfram alpha):
        #   s1 = 1/4 (-sqrt(16 a^2 - 8 a (b + 1) + (b - 1)^2) + 4 a - b + 1)
        s1 = int(
            0.25
            * (-math.sqrt(16 * a**2 - 8 * a * (b + 1) + (b - 1) ** 2) + 4 * a - b + 1)
        )
        s0 = a // s1
        if s0 * s1 != a:
            raise ValueError("Could not find valid split_size_0 and split_size_1")
        # since we don't know which is which, we'll just assume split_size_0 <= split_size_1
        split_size_0 = min(s0, s1)
        split_size_1 = max(s0, s1)

        model = CRAFT(
            in_chans=in_chans,
            window_size=window_size,
            embed_dim=embed_dim,
            depths=depths,
            num_heads=num_heads,
            split_size_0=split_size_0,
            split_size_1=split_size_1,
            mlp_ratio=mlp_ratio,
            qkv_bias=qkv_bias,
            qk_scale=qk_scale,
            upscale=upscale,
            img_range=img_range,
            resi_connection=resi_connection,
        )

        tags = [
            f"{split_size_0}x{split_size_1}",
            f"{embed_dim}dim",
            f"{resi_connection}",
        ]

        return ImageModelDescriptor(
            model,
            state_dict,
            architecture=self,
            purpose="Restoration" if upscale == 1 else "SR",
            tags=tags,
            supports_half=True,  # TODO: Not thoroughly tested
            supports_bfloat16=True,
            scale=upscale,
            input_channels=in_chans,
            output_channels=in_chans,
            size_requirements=SizeRequirements(minimum=16, multiple_of=16),
        )
