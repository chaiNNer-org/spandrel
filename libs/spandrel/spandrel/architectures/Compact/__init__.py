from typing_extensions import override

from spandrel.util import KeyCondition, get_scale_and_output_channels, get_seq_len

from ...__helpers.model_descriptor import Architecture, ImageModelDescriptor, StateDict
from .arch.SRVGG import SRVGGNetCompact


class CompactArch(Architecture[SRVGGNetCompact]):
    def __init__(
        self,
    ) -> None:
        super().__init__(
            id="Compact",
            name="RealESRGAN Compact",
            detect=KeyCondition.has_all(
                "body.0.weight",
                "body.1.weight",
            ),
        )

    @override
    def load(self, state_dict: StateDict) -> ImageModelDescriptor[SRVGGNetCompact]:
        state = state_dict

        highest_num = get_seq_len(state, "body") - 1

        in_nc = state["body.0.weight"].shape[1]
        num_feat = state["body.0.weight"].shape[0]
        num_conv = (highest_num - 2) // 2

        pixelshuffle_shape = state[f"body.{highest_num}.bias"].shape[0]
        scale, out_nc = get_scale_and_output_channels(pixelshuffle_shape, in_nc)

        model = SRVGGNetCompact(
            num_in_ch=in_nc,
            num_out_ch=out_nc,
            num_feat=num_feat,
            num_conv=num_conv,
            upscale=scale,
        )

        return ImageModelDescriptor(
            model,
            state,
            architecture=self,
            purpose="Restoration" if scale == 1 else "SR",
            tags=[f"{num_feat}nf", f"{num_conv}nc"],
            supports_half=True,
            supports_bfloat16=True,
            scale=scale,
            input_channels=in_nc,
            output_channels=out_nc,
        )
