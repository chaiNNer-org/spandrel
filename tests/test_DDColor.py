from spandrel_extra_arches.architectures.DDColor import DDColor, DDColorArch

from .util import (
    ModelFile,
    TestImage,
    assert_image_inference,
    assert_loads_correctly,
    disallowed_props,
    skip_if_unchanged,
)

skip_if_unchanged(__file__)


def test_load():
    assert_loads_correctly(
        DDColorArch(),
        lambda: DDColor(
            nf=64,
            num_queries=20,
            input_size=(32, 32),
            encoder_name="convnext-l",
            decoder_name="MultiScaleColorDecoder",
            num_output_channels=3,
            last_norm="Weight",
            num_scales=3,
            dec_layers=9,
        ),
        lambda: DDColor(
            nf=64,
            num_queries=20,
            input_size=(32, 32),
            encoder_name="convnext-b",
            decoder_name="MultiScaleColorDecoder",
            num_output_channels=2,
            last_norm="Spectral",
            num_scales=3,
            dec_layers=9,
        ),
        lambda: DDColor(
            nf=64,
            num_queries=20,
            input_size=(32, 32),
            encoder_name="convnext-s",
            decoder_name="MultiScaleColorDecoder",
            num_output_channels=2,
            last_norm="Batch",
            num_scales=3,
            dec_layers=9,
        ),
        lambda: DDColor(
            nf=64,
            num_queries=20,
            input_size=(32, 32),
            encoder_name="convnext-t",
            decoder_name="SingleColorDecoder",
            num_output_channels=2,
            last_norm="Spectral",
            num_scales=3,
            dec_layers=9,
        ),
        ignore_parameters={"input_size"},
    )


def test_DDColor_paper_tiny(snapshot):
    file = ModelFile.from_url(
        "https://huggingface.co/piddnad/DDColor-models/resolve/main/ddcolor_paper_tiny.pth"
    )
    model = file.load_model()
    assert model == snapshot(exclude=disallowed_props)
    assert isinstance(model.model, DDColor)
    assert_image_inference(
        file,
        model,
        [TestImage.GRAY_EINSTEIN],
    )
