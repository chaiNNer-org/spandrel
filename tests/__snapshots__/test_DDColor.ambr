# serializer version: 1
# name: test_DDColor_paper_tiny
  ImageModelDescriptor(
    architecture=DDColorArch(
      id='DDColor',
      name='DDColor',
    ),
    input_channels=1,
    output_channels=3,
    purpose='Restoration',
    scale=1,
    size_requirements=SizeRequirements(minimum=0, multiple_of=1, square=False),
    supports_bfloat16=True,
    supports_half=False,
    tags=list([
      'MultiScaleColorDecoder',
      'Spectral',
      'convnext-t',
      '512nf',
      '100nq 3ns 9dl',
    ]),
    tiling=<ModelTiling.INTERNAL: 3>,
  )
# ---
